### PR TITLE
feat: maskbox whitelist

### PR DESCRIPTION
**Pull request by Randolph314**

### CODE REVIEW

File: src/api/merkleProof.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+export async function getMerkleProof(leaf: string, root: string) {
+  try {
+    const response = await fetch(
+      `${MERKLE_PROOF_ENDPOINT}?leaf=${leaf}&root=${root.replace(/^0x/, '')}`,
+    );
+    return (await response.json()) as { proof?: string[]; message?: string; module?: string };
+  } catch (err: any) {
+    throw new Error(err?.message);
+  }
```
Comment (UncleBill) [inline src/api/merkleProof.ts:11]: No need to wrap a try-catch

File: src/page-components/useLocales/zh-CN.json [hunk 2]
```diff
@@ -27,6 +27,7 @@
   "Draw ( {price}/Time )": "抽取（{price}/次）",
   "Lucky Draw": "幸运抽取",
   "Get your unique card (NFT) by lucky draw": "通过幸运抽取获得您唯一的 NFT",
+  "Current address is not in the whitelist.": "当前地址不在白名单内",
```
Comment (UncleBill) [inline src/page-components/useLocales/zh-CN.json:30]: Add translations via crowdin except English version.
https://crowdin.com/translate/mask-network/167/en-zhcn?filter=basic&value=0

File: src/pages/Details/index.tsx [hunk 3]
```diff
@@ -99,6 +106,34 @@ export const Details: FC = memo(() => {
...
+            setIsWhitelisted();
+          }
+          const abiCoder = new ethers.utils.AbiCoder();
+          setProof(abiCoder.encode(['bytes32[]'], [data?.proof?.map((p) => '0x' + p)]));
+        });
+      } catch (error) {
+        setIsNotFetchingProof();
+      }
+    }
+  }, [qualification, account]);
```
Comment (UncleBill) [inline src/pages/Details/index.tsx:135]: Could you rewrite these new hooks into a custom one?

File: src/pages/Details/index.tsx [hunk 4]
```diff
@@ -186,7 +223,8 @@ export const Details: FC = memo(() => {
       {payment && (
         <BuyBox
           open={buyBoxVisible}
-          boxId={boxId}
+          boxId={boxId as string}
```
Comment (UncleBill) [inline src/pages/Details/index.tsx:226]:
```suggestion
          boxId={boxId!}
```

File: src/pages/Edit/Meta.tsx [hunk 5]
```diff
@@ -330,17 +430,65 @@ export const Meta: FC = () => {
...
+          <div className={styles.commonText}>
+            {whitelistNumber > 0 && `total: ${whitelistNumber}`}
+          </div>
+        </div>
       </Field>
 
+      <div className={styles.commonText}>
+        *Addresses should be separated by commas in English (half-width) or space. In Excel, one
+        cell can only be filled with one address, max to 1000.
+      </div>
```
Comment (UncleBill) [inline src/pages/Edit/Meta.tsx:475]: i18n.

Tip: use a short label instead.

File: src/pages/Edit/Meta.tsx [hunk 6]
```diff
@@ -330,17 +430,65 @@ export const Meta: FC = () => {
...
+        onStartUpload={setUploading}
+        onUploaded={handleUploaded}
+        onError={setUploadError}
+      />
+      {uploadError && <div className={styles.uploadWarning}>{uploadError?.message}</div>}
+
+      <div className={styles.commonText}>
+        *Addresses should be separated by commas in English (half-width) or space. In Excel, one
+        cell can only be filled with one address.
+      </div>
```
Comment (UncleBill) [inline src/pages/Edit/Meta.tsx:490]: i18n

File: src/page-components/Maskbox/index.tsx [hunk 7]
```diff
@@ -213,8 +214,8 @@ export const Maskbox: FC<MaskboxProps> = ({
           chainId={chainId!}
           boxId={boxId!}
           boxName={box.name ?? ''}
-          qualification={
-            qualification || '0x0000000000000000000000000000000000000000000000000000000000000000'
+          rootHash={
+            rootHash || '0x0000000000000000000000000000000000000000000000000000000000000000'
```
Comment (UncleBill) [inline src/page-components/Maskbox/index.tsx:218]: You already check it above.

File: src/components/UploadButton/index.tsx [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+const getMaxSize = (mediaType: MediaType) => {
+  switch (mediaType) {
+    case MediaType.Video:
+    case MediaType.Audio:
+      return 30 * 1024 * 1024;
+    case MediaType.Image:
+    case MediaType.Unknown:
+      return 2 * 1024 * 1024;
+    case MediaType.csv:
```
Comment (UncleBill) [inline src/components/UploadButton/index.tsx:26]: Capitalized first letter for enum key.
```suggestion
    case MediaType.Csv:
```

**Review by UncleBill — APPROVED** [review]

**Review by UncleBill — APPROVED** [review]

File: src/abi/Maskbox.json [hunk 9]
```diff
@@ -311,6 +311,11 @@
         "internalType": "uint256",
```
Comment (guanbinrui) [inline src/abi/Maskbox.json:311]: `MaskBox.json`

File: src/components/UploadButton/index.tsx [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+const getMaxSize = (mediaType: MediaType) => {
+  switch (mediaType) {
+    case MediaType.Video:
+    case MediaType.Audio:
+      return 30 * 1024 * 1024;
+    case MediaType.Image:
+    case MediaType.Unknown:
+      return 2 * 1024 * 1024;
+    case MediaType.Csv:
+      return 10 * 1024 * 1024;
```
Comment (guanbinrui) [inline src/components/UploadButton/index.tsx:27]: Why doesn't it provide a default size?

Comment (guanbinrui) [inline src/components/UploadButton/index.tsx:27]: Why doesn't it provide a default size?

File: src/components/UploadButton/index.tsx [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+          const fileReader = new FileReader();
+          fileReader.readAsText(f);
+          fileReader.onload = function () {
+            const addressList = this.result?.toString()?.split(/\s+/);
+            if (onUploaded) {
+              onUploaded({
+                fileAddressList: addressList,
+                whitelistFileName: f.name,
+              });
+            }
```
Comment (guanbinrui) [inline src/components/UploadButton/index.tsx:66]:
```
onUploaded?.({
  fileAddressList: addressList,
  whitelistFileName: f.name,
});
```

File: src/components/UploadButton/index.tsx [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+  const [invalidMessage, setInvalidMessage] = useState('');
+  const [dragingIn, setDragingIn] = useState(false);
+  const handleUpload = async (file?: File) => {
+    setInvalidMessage('');
+    try {
+      onStartUpload?.();
+      await upload(file, (f) => {
+        const mediaType = getMediaType(f.name);
+        const maxSize = getMaxSize(mediaType);
+        if (f.size > maxSize) {
```
Comment (guanbinrui) [inline src/components/UploadButton/index.tsx:52]: As your current implementation, the `maxSize` could be `undefined`.

File: src/components/UploadButton/index.tsx [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+          throw new Error(t('File size up to {size}MB', { size: maxSize / (1024 * 1024) }));
+        } else if (!f.name.toLowerCase().match(FILE_EXT_RE)) {
+          throw new Error(t('You should upload a file, only accept csv'));
+        } else {
+          const fileReader = new FileReader();
+          fileReader.readAsText(f);
+          fileReader.onload = function () {
```
Comment (guanbinrui) [inline src/components/UploadButton/index.tsx:59]: You should handle `onerror`.

File: src/lib/constants/contracts.ts [hunk 14]
```diff
@@ -10,11 +10,17 @@ export const chainShortNameMap: Record<number, string> = {
 export const contractAddresses: Record<string, Record<string, string>> = {
   Mainnet: {
     Maskbox: '0x294428f04b0F9EbC49B7Ad61E2736ebD6808c145',
```
Comment (guanbinrui) [inline src/lib/constants/contracts.ts:12]: `MaskBox`

File: src/page-components/BuyBox/useOpenBox.tsx [hunk 15]
```diff
@@ -23,86 +23,89 @@ export function useOpenBox(
...
+    async (qualification: string) => {
+      if (!contract || !ethersProvider || !account || !chainId) return;
+      setLoading(true);
+      showToast({
+        title: 'Drawing',
+        message: t('Sending transaction'),
       });
+      try {
+        const purchasedNfts = await getPurchasedNft(boxId, account);
+        // TODO the proof parameter
```
Comment (guanbinrui) [inline src/page-components/BuyBox/useOpenBox.tsx:36]: Do you forget to clean this TODO?

File: src/page-components/BuyBox/useOpenBox.tsx [hunk 16]
```diff
@@ -23,86 +23,89 @@ export function useOpenBox(
...
+              </a>
+            </span>
+          ),
+        });
+        await tx.wait(1);
+        await ethersProvider.getLogs(contract.filters.OpenSuccess());
+        const newlyPurchasedNfts = (await getPurchasedNft(boxId, account)).filter(
+          (id) => !purchasedNfts.includes(id),
+        );
+        console.log({ newlyPurchasedNfts });
```
Comment (guanbinrui) [inline src/page-components/BuyBox/useOpenBox.tsx:72]:
```suggestion
```

File: src/pages/Details/index.tsx [hunk 17]
```diff
@@ -25,16 +26,20 @@ export const Details: FC = memo(() => {
...
+    const rootHash = params.get('rootHash') || undefined;
     return {
       chainId: chainId ? parseInt(chainId, 10) : null,
       boxId,
+      rootHash,
     };
   }, [search]);
 
+  const { isWhitelisted, isFetchingProof, proof } = useMerkleProof(rootHash);
+  console.log({ isWhitelisted, isFetchingProof, proof });
```
Comment (guanbinrui) [inline src/pages/Details/index.tsx:42]:
```suggestion
```

Comment (guanbinrui) [inline src/pages/Details/index.tsx:42]:
```suggestion
```

File: src/pages/Edit/Meta.tsx [hunk 18]
```diff
@@ -103,7 +143,34 @@ export const Meta: FC = () => {
...
+            address
+              ?.replace(/0x/, '')
+              ?.match(/.{1,2}/g)
+              ?.map((byte) => parseInt(byte, 16)),
+          )
+          .filter((numbers) => numbers && numbers.length)
+          .map((numbers) => new Uint8Array(numbers as number[]))
+          .map((uint8Array) => Buffer.from(uint8Array).toString('base64'));
+        const res = await getMerkleProof(leaves);
+        formData.merkleProof = '0x' + res?.root;
```
Comment (guanbinrui) [inline src/pages/Edit/Meta.tsx:166]: `'0x' + undefined`= `0xundefined`

Comment (guanbinrui) [inline src/pages/Edit/Meta.tsx:166]: `'0x' + undefined`= `0xundefined`

File: src/pages/Details/useMerkleProof.ts [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+  const [isFetchingProof, setIsFetchingProof, setIsNotFetchingProof] = useBoolean(true);
+  const [isWhitelisted, setIsWhitelisted, setNotWhitelisted] = useBoolean(true);
+  const [proof, setProof] = useState<string>();
+
+  useEffect(() => {
+    if (rootHash && account) {
+      const leafArray = account
+        ?.replace(/0x/, '')
+        ?.match(/.{1,2}/g)
+        ?.map((byte) => parseInt(byte, 16));
```
Comment (guanbinrui) [inline src/pages/Details/useMerkleProof.ts:18]:
```suggestion
        ?.map((byte) => Number.parseInt(byte, 16));
```

File: src/pages/Details/useMerkleProof.ts [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+      const leaf = encodeURIComponent(
+        Buffer.from(new Uint8Array(leafArray as number[])).toString('base64'),
+      );
+      setIsFetchingProof();
+      getMerkleProof(leaf as string, rootHash?.replace(/0x/, ''))
```
Comment (guanbinrui) [inline src/pages/Details/useMerkleProof.ts:23]:
```suggestion
      getMerkleProof(leaf as string, rootHash?.replace(/^0x/, ''))
```

File: src/pages/Edit/hooks/useCreateMaskbox.ts [hunk 21]
```diff
@@ -39,11 +40,15 @@ export function useCreateMaskbox() {
       endTime,
       formData.sellAll,
       formData.selectedNFTIds,
-      formData.whiteList || ZERO_ADDRESS,
+      formData.merkleProof === '0x0000000000000000000000000000000000000000000000000000000000000000'
+        ? ZERO_ADDRESS
+        : merkleTreeAddress,
+      //formData.whiteList || ZERO_ADDRESS,
```
Comment (guanbinrui) [inline src/pages/Edit/hooks/useCreateMaskbox.ts:46]:
```suggestion
```

File: src/pages/Edit/Meta.tsx [hunk 22]
```diff
@@ -330,17 +430,56 @@ export const Meta: FC = () => {
...
+              className={styles.button}
+              disabled={isEditting || Boolean(formData?.whitelistFileName)}
+              colorScheme="primary"
+              onClick={editwhitelist}
+            >
+              {t('Edit')}
+            </Button>
+          </div>
+          <div className={styles.commonText}>
+            {whitelistNumber > 0 && `total: ${whitelistNumber}`}
```
Comment (guanbinrui) [inline src/pages/Edit/Meta.tsx:464]: `t()`

File: src/pages/Edit/hooks/useCreateMaskbox.ts [hunk 23]
```diff
@@ -39,11 +40,15 @@ export function useCreateMaskbox() {
       endTime,
       formData.sellAll,
       formData.selectedNFTIds,
-      formData.whiteList || ZERO_ADDRESS,
+      formData.merkleProof === '0x0000000000000000000000000000000000000000000000000000000000000000'
```
Comment (guanbinrui) [inline src/pages/Edit/hooks/useCreateMaskbox.ts:43]: Move `0x0000000000000000000000000000000000000000000000000000000000000000` to be a constant variable.

File: src/pages/Edit/atoms.ts [hunk 24]
```diff
@@ -150,6 +170,26 @@ export const validationsAtom = atom<string[]>((get) => {
   ) {
     validations.push('End date should be later than start date');
   }
+  if (formData?.fileAddressList && formData.fileAddressList?.length > 0) {
+    if (formData?.fileAddressList?.length > 1000)
+      validations.push('Please limit whitelist address number less than 1000');
```
Comment (guanbinrui) [inline src/pages/Edit/atoms.ts:175]: These messages should have transactions.

File: src/pages/Edit/Meta.tsx [hunk 25]
```diff
@@ -103,7 +143,34 @@ export const Meta: FC = () => {
...
+          ? formData?.whitelist?.split(',')
+          : undefined;
+
+      if (whitelist) {
+        const leaves = whitelist
+          .map((address) =>
+            address
+              ?.replace(/0x/, '')
+              ?.match(/.{1,2}/g)
+              ?.map((byte) => parseInt(byte, 16)),
```
Comment (guanbinrui) [inline src/pages/Edit/Meta.tsx:160]:
```suggestion
              ?.map((byte) => Number.parseInt(byte, 16)),
```

File: src/pages/Edit/Meta.tsx [hunk 26]
```diff
@@ -103,7 +143,34 @@ export const Meta: FC = () => {
...
+          )
+          .filter((numbers) => numbers && numbers.length)
```
Comment (guanbinrui) [inline src/pages/Edit/Meta.tsx:162]:
```suggestion
          .filter((numbers) => numbers?.length)
```

File: src/pages/Edit/Meta.tsx [hunk 27]
```diff
@@ -158,6 +229,35 @@ export const Meta: FC = () => {
...
+    ({
+      fileAddressList,
+      whitelistFileName,
+    }: {
+      fileAddressList?: string[];
+      whitelistFileName: string;
+    }) => {
+      setUploadError(null);
+      setNotUploading();
+      if (fileAddressList && fileAddressList?.length > 1000) {
```
Comment (guanbinrui) [inline src/pages/Edit/Meta.tsx:250]:
```suggestion
      if (fileAddressList && fileAddressList.length > 1000) {
```

File: src/pages/Edit/Meta.tsx [hunk 28]
```diff
@@ -428,6 +567,12 @@ export const Meta: FC = () => {
         open={confirmDialogVisible}
         tokens={formData.sellAll ? ownedERC721Tokens : selectedERC721Tokens}
         onClose={closeConfirmDialog}
+        whitelistAddressList={
+          formData?.whitelist && formData?.whitelist?.length > 0
+            ? formData?.whitelist?.split(',')
```
Comment (guanbinrui) [inline src/pages/Edit/Meta.tsx:572]:
```suggestion
            ? formData?.whitelist?.split(/\s*,\s*/)
```

File: src/pages/Edit/atoms.ts [hunk 29]
```diff
@@ -82,8 +85,25 @@ export const metaFullfilledAtom = atom((get) => {
     formData.endAt &&
     new Date(formData.endAt).getTime() > new Date(formData.startAt).getTime();
   const limitIsOk = formData.limit && formData.limit > 0 && formData.limit < 256;
+  const whitelist = formData?.whitelist;
+  const fileAddressList = formData?.fileAddressList;
+  const whitelistIsOk =
+    !whitelist ||
+    (whitelist?.split(',')?.length <= 1000 &&
```
Comment (guanbinrui) [inline src/pages/Edit/atoms.ts:92]: Please reduce duplicate operations.

File: src/pages/Edit/atoms.ts [hunk 30]
```diff
@@ -82,8 +85,25 @@ export const metaFullfilledAtom = atom((get) => {
...
+      whitelist?.split(',')?.every((address) => /^(0x)?[0-9a-zA-Z]{40}$/.test(address)));
+  const fileAddressListIsOk =
+    !fileAddressList ||
+    (fileAddressList?.length <= 1000 &&
```
Comment (guanbinrui) [inline src/pages/Edit/atoms.ts:96]:
```suggestion
    (fileAddressList.length <= 1000 &&
```

File: src/pages/Edit/atoms.ts [hunk 31]
```diff
@@ -82,8 +85,25 @@ export const metaFullfilledAtom = atom((get) => {
...
+      fileAddressList?.every((address) => /^(0x)?[0-9a-zA-Z]{40}$/.test(address)));
```
Comment (guanbinrui) [inline src/pages/Edit/atoms.ts:97]:
```suggestion
    (fileAddressList.length <= 1000 &&
      fileAddressList.every((address) => /^(0x)?[0-9a-zA-Z]{40}$/.test(address)));
```

File: src/pages/Edit/atoms.ts [hunk 32]
```diff
@@ -150,6 +170,26 @@ export const validationsAtom = atom<string[]>((get) => {
...
+  if (formData?.fileAddressList && formData.fileAddressList?.length > 0) {
+    if (formData?.fileAddressList?.length > 1000)
+      validations.push('Please limit whitelist address number less than 1000');
+    if (
+      formData.fileAddressList?.some((address) => /^(0x)?[0-9a-zA-Z]{40}$/.test(address) === false)
+    ) {
+      validations.push('Please input or upload correct address');
+    }
+  }
+  if (formData?.whitelist && formData.whitelist?.length > 0) {
```
Comment (guanbinrui) [inline src/pages/Edit/atoms.ts:182]:
```suggestion
  if (formData?.whitelist && formData.whitelist.length > 0) {
```

File: src/pages/Edit/atoms.ts [hunk 33]
```diff
@@ -150,6 +170,26 @@ export const validationsAtom = atom<string[]>((get) => {
...
+    if (formData?.whitelist?.split(',')?.length > 1000)
+      validations.push('Please limit whitelist address number less than 1000');
+    if (
+      formData.whitelist
+        ?.split(',')
+        ?.some((address) => /^(0x)?[0-9a-zA-Z]{40}$/.test(address) === false)
+    ) {
+      validations.push('Please input or upload correct address');
+    }
+  }
```
Comment (guanbinrui) [inline src/pages/Edit/atoms.ts:192]: Please reduce duplicate operations.

File: src/pages/Edit/atoms.ts (hunk 24 above)
Comment (Randolph314) [reply src/pages/Edit/atoms.ts:183]: cannot use i18n in unfunctional component

**Review by UncleBill — APPROVED** [review]

File: src/pages/Details/useMerkleProof.ts [hunk 34]
```diff
@@ -0,0 +1,39 @@
...
+import { ethers } from 'ethers';
+import { useWeb3Context } from '@/contexts';
+import { useBoolean } from '@/utils';
+import { getMerkleProof } from '@/api/merkleProof';
+
+export function useMerkleProof(rootHash?: string) {
+  const { account } = useWeb3Context();
+  const [isFetchingProof, setIsFetchingProof, setIsNotFetchingProof] = useBoolean(true);
+  const [isWhitelisted, setIsWhitelisted, setNotWhitelisted] = useBoolean(true);
+  const [proof, setProof] = useState<string>();
```
Comment (guanbinrui) [inline src/pages/Details/useMerkleProof.ts:11]: Initial with the default proof here.

File: src/pages/Details/useMerkleProof.ts [hunk 35]
```diff
@@ -0,0 +1,39 @@
...
+
+  useEffect(() => {
+    if (rootHash && account) {
+      const leafArray = account
+        ?.replace(/0x/, '')
```
Comment (guanbinrui) [inline src/pages/Details/useMerkleProof.ts:16]:
```suggestion
        ?.replace(/^0x/, '')
```

File: src/pages/Details/useMerkleProof.ts [hunk 36]
```diff
@@ -0,0 +1,39 @@
...
+        ?.match(/.{1,2}/g)
+        ?.map((byte) => Number.parseInt(byte, 16));
+      const leaf = encodeURIComponent(
+        Buffer.from(new Uint8Array(leafArray as number[])).toString('base64'),
```
Comment (guanbinrui) [inline src/pages/Details/useMerkleProof.ts:20]: `leafArray ?? []`

File: src/pages/Details/useMerkleProof.ts [hunk 37]
```diff
@@ -0,0 +1,39 @@
...
+      );
+      setIsFetchingProof();
+      getMerkleProof(leaf as string, rootHash?.replace(/^0x/, ''))
```
Comment (guanbinrui) [inline src/pages/Details/useMerkleProof.ts:23]: redundant type casting

File: src/pages/Details/useMerkleProof.ts [hunk 38]
```diff
@@ -0,0 +1,39 @@
...
+        ?.then((data) => {
+          setIsNotFetchingProof();
+          if (data?.message === 'leaf not found') {
+            setNotWhitelisted();
+          }
+          if (data?.proof) {
+            setIsWhitelisted();
+          }
+          const abiCoder = new ethers.utils.AbiCoder();
+          setProof(abiCoder.encode(['bytes32[]'], [data?.proof?.map((p) => '0x' + p)]));
```
Comment (guanbinrui) [inline src/pages/Details/useMerkleProof.ts:33]:
```suggestion
          setProof(abiCoder.encode(['bytes32[]'], [data?.proof?.map((p) => '0x' + p) ?? []]));
```

File: src/pages/Edit/Meta.tsx [hunk 39]
```diff
@@ -103,7 +142,34 @@ export const Meta: FC = () => {
...
+          ? formData?.fileAddressList
+          : formData?.whitelist && formData?.whitelist?.length > 0
+          ? formData?.whitelist?.split(',')
+          : undefined;
+
+      if (whitelist) {
+        const leaves = whitelist
+          .map((address) =>
+            address
+              ?.replace(/0x/, '')
```
Comment (guanbinrui) [inline src/pages/Edit/Meta.tsx:157]:
```suggestion
              ?.replace(/^0x/, '')
```

File: src/pages/Edit/Meta.tsx [hunk 40]
```diff
@@ -103,7 +142,34 @@ export const Meta: FC = () => {
...
+        formData?.fileAddressList && formData?.fileAddressList?.length > 0
+          ? formData?.fileAddressList
+          : formData?.whitelist && formData?.whitelist?.length > 0
+          ? formData?.whitelist?.split(',')
+          : undefined;
+
+      if (whitelist) {
+        const leaves = whitelist
+          .map((address) =>
+            address
```
Comment (guanbinrui) [inline src/pages/Edit/Meta.tsx:156]: You need a utility for converting an ethereum address into a base64 string.

File: src/pages/Edit/atoms.ts (hunk 24 above)
Comment (guanbinrui) [reply src/pages/Edit/atoms.ts:183]: @UncleBill Any advice?

File: src/pages/Edit/Meta.tsx [hunk 41]
```diff
@@ -78,8 +92,33 @@ export const Meta: FC = () => {
 
   const { saveBox } = useRSS3();
   const [confirmDialogVisible, openConfirmDialog, closeConfirmDialog] = useBoolean();
+  const [whitelistNumber, setwhitelistNumber] = useState<number>(0);
```
Comment (guanbinrui) [inline src/pages/Edit/Meta.tsx:95]:
```suggestion
  const [whitelistNumber, setwhitelistNumber] = useState(0);
```

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**